### PR TITLE
Fix duplicate compaction summaries after chat reload

### PR DESCRIPTION
**Pull request by wwwillchen**
## Summary

Reloading a chat after mid-turn compaction shows the same compaction twice: once inside the assistant reply and again at the bottom as a separate message. Use one display projection for streaming, chat reloads, window bootstrap, and full message refreshes to hide the duplicate.

- Keep the summary row in SQLite for model context. Hide it only when its exact compaction block already appears in an assistant reply from the same triggering user turn with a compatible timestamp; preserve standalone summaries when inline output is missing or differs.
- Build persisted and inline compaction blocks through one helper, including the empty/whitespace-summary fallback. Apply the current streamed response before filtering and remove the separate streaming hidden-ID heuristic.
- Cover empty summaries, streaming/reload consistency, retained database history, exact preservation of standalone messages, and a real mid-turn compaction followed by get-chat.
- Validation limitation: the real mid-turn integration test passes the new UI/reload assertions but fails its existing Git-provenance snapshot ("created commit" versus "created no commit"). The identical failure reproduces on clean upstream main at 1b2874a84; the snapshot is unchanged here. All 161 focused tests pass.

- Share block construction and extraction through the compaction utilities; fixtures use the production builder, and the persistence test asserts exactly one block. Regressions cover identical summaries across turns and missing inline output.

---

> [!NOTE]
> **Medium Risk**
> Changes chat display projection and streaming full-update payloads across multiple IPC entry points; incorrect dedupe logic could hide the only compaction indicator or show duplicates again, but behavior is narrowly scoped to `isCompactionSummary` rows with explicit fallback tests.
> 
> **Overview**
> **Mid-turn compaction** now keeps a separate `isCompactionSummary` row for model history while the UI shows the indicator inline in the assistant turn. Reloading previously surfaced both; this PR centralizes renderer projection in **`toRendererMessages`**, which drops a summary row only when its exact `<dyad-compaction>` block already appears in an earlier assistant message (with compatible id/timestamp). Standalone summaries stay visible when inline output is missing or differs.
> 
> All full-message IPC paths (`get-chat`, stream chunks, window bootstrap) switch from per-row `toRendererMessage` to **`toRendererMessages`**. The local agent removes ad-hoc `hiddenMessageIds` / `getMidTurnCompactionSummaryIds` and applies live placeholder content before filtering in **`sendResponseChunk`**.
> 
> **`buildCompactionBlock`** unifies persisted and inline compaction XML (empty-summary fallback and escaping) in `compaction_handler` and `local_agent_handler`. Docs in `electron-ipc.md` note the pattern. Regression tests cover reload, DB retention, and edge cases that must keep a standalone indicator.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a26e41ae6d95c47ee08e6cddb1637c9a65441c49. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<a href="https://cubic.dev/pr/dyad-sh/dyad/pull/4630?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/ipc/utils/renderer_chat_message.ts [hunk 1]
```diff
@@ -20,13 +20,43 @@ export const rendererMessageColumns = {
...
+      )?.[0];
+      // Check the actual inline block: timestamps alone could hide the only
+      // indicator if compaction completed but the turn was never persisted.
+      return (
+        !block ||
+        !inlineMessages.some(
+          (inline) =>
+            inline.id < row.id &&
+            inline.createdAt.getTime() <= row.createdAt.getTime() &&
+            inline.content.includes(block),
```
[the remark this answers is not in the document]
Comment (wwwillchen) [reply src/ipc/utils/renderer_chat_message.ts]: Addressed in a26e41ae6. Both persisted and inline summaries now use buildCompactionBlock, including the empty/whitespace fallback and XML escaping. Tests cover the real persistence producer, streaming with empty summaries, and real mid-turn compaction followed by get-chat. The new UI/reload assertions pass; a later, unrelated Git-provenance snapshot failure reproduces identically on clean upstream main.

File: src/ipc/utils/renderer_chat_message.ts [hunk 2]
```diff
@@ -20,13 +20,43 @@ export const rendererMessageColumns = {
...
+  isCompactionSummary: true,
 } as const;
 
 export type RendererMessageRow = Pick<
   typeof messages.$inferSelect,
   keyof typeof rendererMessageColumns
 >;
 
+/** Keep model-history summaries in the DB, but do not display them twice. */
+export function toRendererMessages(rows: RendererMessageRow[]): Message[] {
```
[the remark this answers is not in the document]
Comment (wwwillchen) [reply src/ipc/utils/renderer_chat_message.ts:33]: Addressed in a26e41ae6. Removed getMidTurnCompactionSummaryIds and hiddenMessageIdsForStreaming. Full streaming updates now apply live placeholder content before calling the same toRendererMessages projection used by reload and window bootstrap. Regression coverage checks streaming/reload message IDs for normal, empty, and whitespace summaries.

File: src/ipc/utils/renderer_chat_message.ts [hunk 3]
```diff
@@ -20,13 +20,43 @@ export const rendererMessageColumns = {
...
+    .filter((row) => {
+      if (!row.isCompactionSummary) return true;
+      const block = row.content.match(
+        /<dyad-compaction\b[^>]*>[\s\S]*?<\/dyad-compaction>/,
+      )?.[0];
+      // Check the actual inline block: timestamps alone could hide the only
+      // indicator if compaction completed but the turn was never persisted.
+      return (
+        !block ||
+        !inlineMessages.some(
```
[the remark this answers is not in the document]
Comment (wwwillchen) [reply src/ipc/utils/renderer_chat_message.ts:58]: Addressed in a26e41ae6. Both persisted and inline summaries now use buildCompactionBlock, including the empty/whitespace fallback and XML escaping. Tests cover the real persistence producer, streaming with empty summaries, and real mid-turn compaction followed by get-chat. The new UI/reload assertions pass; a later, unrelated Git-provenance snapshot failure reproduces identically on clean upstream main.

File: src/ipc/utils/renderer_chat_message.test.ts [hunk 4]
```diff
@@ -18,3 +19,58 @@ it("projects the submission identity so history can precede acceptance", () => {
...
+  ],
+])("preserves a standalone summary with %s", (_, overrides) => {
+  const reply = row(overrides);
+  const summary = row({
+    id: 2,
+    content: compaction,
+    isCompactionSummary: true,
+    createdAt: new Date(2000),
+  });
+  expect(toRendererMessages([reply, summary])).toHaveLength(2);
```
[the remark this answers is not in the document]
Comment (wwwillchen) [reply src/ipc/utils/renderer_chat_message.test.ts]: Addressed in a26e41ae6. Each preservation case now asserts both returned messages explicitly, including their IDs and contents, rather than only checking the output length.

**Comment by wwwillchen** [comment]
## 🤖 Codex Review Summary

### PR Confidence: 4/5
All four trusted review threads are addressed and explicitly resolved; 147 focused tests and formatting/lint/type checks pass. The new real-compaction UI/reload assertions pass, but the existing Git-provenance snapshot fails identically on clean upstream main (1b2874a84). Fresh CI/review results are still pending.

### Unresolved Threads

_No unresolved threads at the post-push audit._

### Resolved Threads

| Issue | Rationale | Link |
|-------|-----------|------|
| Empty/whitespace summaries diverge between producers (two reports) | Both producers use a shared compaction-block builder with the same fallback and escaping; real persistence and streaming tests cover these cases. | [Review 1](https://github.com/dyad-sh/dyad/pull/4630#discussion_r4050463840), [Review 2](https://github.com/dyad-sh/dyad/pull/4630#discussion_r4050465421) |
| Streaming and reload use separate display rules | Removed the streaming-only hidden-ID heuristic; streaming applies live placeholder content before using the shared display projection. | [View](https://github.com/dyad-sh/dyad/pull/4630#discussion_r4050463847) |
| Preservation tests only check length | Assertions now compare both projected messages, preserving identity and content. | [View](https://github.com/dyad-sh/dyad/pull/4630#discussion_r4050465431) |

Pushed commit: a26e41ae6. No untrusted commenters or items requiring human attention were encountered. No product-principle judgment calls were needed.

---
🤖 Generated by Codex

File: src/ipc/utils/renderer_chat_message.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (wwwillchen) [reply src/ipc/utils/renderer_chat_message.ts]: Fixed in 7ac89bdb5. Each summary now finds its triggering user by insertion ID and only matches inline blocks in assistant rows from that turn. It preserves the summary if the triggering user is absent. Regressions cover identical empty and nonempty summaries across turns, backdated pre-turn summaries, lost inline output, and valid same-turn deduplication. All 161 targeted tests pass.

File: src/ipc/utils/renderer_chat_message.ts [hunk 5]
```diff
@@ -20,13 +20,43 @@ export const rendererMessageColumns = {
...
+    (row) =>
+      row.role === "assistant" &&
+      !row.isCompactionSummary &&
+      row.content.includes("<dyad-compaction"),
+  );
+  return rows
+    .filter((row) => {
+      if (!row.isCompactionSummary) return true;
+      const block = row.content.match(
+        /<dyad-compaction\b[^>]*>[\s\S]*?<\/dyad-compaction>/,
```
[the remark this answers is not in the document]
Comment (wwwillchen) [reply src/ipc/utils/renderer_chat_message.ts]: Implemented in 7ac89bdb5. COMPACTION_TAG and extractCompactionBlocks now live beside buildCompactionBlock; the display projection no longer embeds the tag name or regex. Projection and chat-handler fixtures use the production builder. Parser coverage includes multiple blocks, escaped tag text, and incomplete blocks. The real mid-turn compaction/reload integration test also exercises the producers together.

File: src/ipc/handlers/compaction/compaction_handler.integration.test.ts [hunk 6]
```diff
@@ -140,6 +141,28 @@ describe("performCompaction", () => {
...
+        chatId,
+        "/tmp/test-app",
+        "request-id",
+        undefined,
+        { createdAtStrategy: "now" },
+      );
+      expect(result.success).toBe(true);
+      const summaries = await loadSummaryMessages();
+      expect(summaries).toHaveLength(1);
+      expect(summaries[0].content).toContain(buildCompactionBlock(summary));
```
[the remark this answers is not in the document]
Comment (wwwillchen) [reply src/ipc/handlers/compaction/compaction_handler.integration.test.ts]: Implemented in 7ac89bdb5: the persistence test now asserts exactly one occurrence of the canonical block. The original UI bug concerned duplicate message rows, which is covered separately, but this additional assertion strengthens the producer test.

File: src/ipc/handlers/compaction/compaction_utils.ts [hunk 7]
```diff
@@ -1,6 +1,13 @@
 /**
  * Shared utilities for context compaction.
  */
+import { escapeXmlContent } from "../../../../shared/xmlEscape";
+
+/** Both persisted summaries and inline indicators must use the same block. */
+export function buildCompactionBlock(summary?: string): string {
+  const text = summary?.trim() ? summary : "Conversation compacted.";
```
[the remark this answers is not in the document]
Comment (wwwillchen) [reply src/ipc/handlers/compaction/compaction_utils.ts:10]: Fixed in 7ac89bdb5. Each summary now finds its triggering user by insertion ID and only matches inline blocks in assistant rows from that turn. It preserves the summary if the triggering user is absent. Regressions cover identical empty and nonempty summaries across turns, backdated pre-turn summaries, lost inline output, and valid same-turn deduplication. All 161 targeted tests pass.

**Comment by wwwillchen** [comment]
## 🤖 Codex Review Summary

### PR Confidence: 4/5
Fixed the cross-turn matching bug and implemented both review suggestions in 7ac89bdb5. All 161 targeted tests plus formatting, lint, and type checks pass. Fresh CI results remain pending; the previously documented real-compaction Git-provenance snapshot mismatch reproduces on clean main.

### Unresolved Threads

_No unresolved threads at the post-push audit._

### Resolved Threads

| Issue | Rationale | Link |
|-------|-----------|------|
| Identical summaries across turns hide a real indicator (two reports) | Match only assistant messages after the summary's triggering user and before the summary by insertion ID. Cover empty/nonempty repeated text, pre-turn summaries, and lost inline output. | [Report 1](https://github.com/dyad-sh/dyad/pull/4630#discussion_r4050690315), [Report 2](https://github.com/dyad-sh/dyad/pull/4630#discussion_r4050691194) |
| Block parsing duplicates producer format | Shared tag constant and extraction helper; fixtures use the production builder. | [View](https://github.com/dyad-sh/dyad/pull/4630#discussion_r4050690318) |
| Persistence assertion only checks containment | Assert the canonical block occurs exactly once. | [View](https://github.com/dyad-sh/dyad/pull/4630#discussion_r4050691182) |

No human-attention flags or untrusted commenters were encountered.

---
🤖 Generated by Codex
